sound/soc/sof/core.c

```diff
@@ -310,6 +310,7 @@ int snd_sof_device_probe(struct device *dev, struct snd_sof_pdata *plat_data)
 	/* check all mandatory ops */
 	if (!sof_ops(sdev) || !sof_ops(sdev)->probe || !sof_ops(sdev)->run ||
 	    !sof_ops(sdev)->block_read || !sof_ops(sdev)->block_write ||
+	    !sof_ops(sdev)->mailbox_read || !sof_ops(sdev)->mailbox_write ||
```

Comment:
nitpick: it is mandatory for platforms which will use the generic functions from the new stream-ipc.c
HDA will not be using it.

Reply:
Oh, I see. This complicates a little bit the code. Should I add this under SND_SOC_SOF_INTEL_HIFI_EP_IPC ?

Reply:
@ujfalusi I spent some time to look at this. I don't understand why do you say that HDA will not be using mailbox operations?

For example I have this:

```
intel/hda-ipc.c:                sof_mailbox_read(sdev, sdev->dsp_box.offset, p, sz);
intel/hda-ipc.c:                sof_mailbox_read(sdev, hda_stream->stream.posn_offset, p, sz);
```

Reply:
@dbaluta, sorry for the confusion, what I meant is that HDA will not need to use the new ops as it is not using the generic platform ipc code you are moving out (from sof/intel) to be used by imx as well.
All platforms used the sof_mailbox_* functions directly from sof/utils.c, there were no need for callback abstraction.
With a generic and platform independent code no at the core, we need to use ops.

There is the option to do something like this (I don't have the code at hand):
```
int sof_mailbox_read(sdev, offset, p, sz)
{
	if (sdev->platform_cb->mailbox_read)
		return sdev->platform_cb->mailbox_read(sdev, offset, p, sz);
...
	/* the current implementation */
	return 0;
} 
```

We could convert other ops to similar one if it makes sense.

I would rather avoid using the callback way from platform code since it makes no sense to just add another layer of obfuscation and call depth. The platform code knows how it accesses to mailbox, how it does block copy, how it accesses registers (read/write), etc.

Reply:
I see your point. Can you help me identify HDA platforms:

```
intel/bdw.c:    .mailbox_read   = sof_mailbox_read,
intel/byt.c:    .mailbox_read   = sof_mailbox_read,
intel/cnl.c:    .mailbox_read   = sof_mailbox_read,
intel/pci-tng.c:        .mailbox_read   = sof_mailbox_read,
intel/tgl.c:    .mailbox_read   = sof_mailbox_read,
```

Reply:
@dbaluta, I would see if the other ops got assigned to something hda_* , if they are then it is hda platform.
Sorry, I'm new at Intel and I still have gaps to cover...

Reply:
@ujfalusi  @dbaluta We basicly have two different implementations for this for Intel hardware, intel-ipc.c (older, thus the name) and hda-ipc.c (new, more descriptive name).

Wouldn't be easier to just have this as mandatory. Even if some of the Intel platforms wouldn't use the mailbox read via ops, they'll still set the ops (and can use the generic implementation)...?